database/database.go

```diff
@@ -0,0 +1,232 @@
+	createString := string(creates)
+	if err != nil {
+		panic(err)
+	}
+
+	conn, err := sqlite3.Open(name)
+	defer conn.Close()
+	if err != nil {
+		panic(err)
+	}
+	conn.Exec(createString)
+
+	return Database{name}
+}
+
+// CreateEmptyFile creates a new file with no contents, gives ownership to userName
+// Returns the new file's ID
+func (db *Database) CreateEmptyFile(fileName string, userID int) (int, int, error) {
+	conn, err := sqlite3.Open(db.Path)
+	defer conn.Close()
+	if err != nil {
+		return -1, -1, err
+	}
+	err = conn.Exec("PRAGMA foreign_keys = ON")
+	if err != nil {
+		return -1, -1, err
+	}
+
+	statement, err := conn.Prepare("INSERT INTO files (filename, owner) VALUES(?, ?)")
+	defer statement.Close()
+	if err != nil {
+		return -1, -1, err
+	}
+	statement.Exec(fileName, userID)
+
+	// Get file ID
+	rows, err := conn.Query("SELECT LAST_INSERT_ROWID();")
+	if rows != nil {
+		defer rows.Close()
+	}
+	if err != nil {
+		return -1, -1, err
+	}
+	var fileID int
+	rows.Scan(&fileID)
+
+	if fileID == 0 {
+		err = fmt.Errorf("invalid user ID %v", userID)
+		return -1, -1, err
+	}
+
+	return fileID, 1, nil
+}
+
+// CreateUser creates a user with the given username
+// returns the new user's ID
+func (db *Database) CreateUser(username string) (int, error) {
+	conn, err := sqlite3.Open(db.Path)
+	defer conn.Close()
+	if err != nil {
+		return -1, err
+	}
+	conn.Exec("PRAGMA foreign_keys = ON")
+
+	statement, err := conn.Prepare("INSERT INTO users (username) VALUES(?)")
+	if err != nil {
+		return -1, err
+	}
+
+	statement.Exec(username)
+
+	// Get user ID
+	rows, err := conn.Query("SELECT LAST_INSERT_ROWID();")
+	if rows != nil {
+		defer rows.Close()
+	}
+	if err != nil {
+		return -1, err
+	}
+	var userID int
+	rows.Scan(&userID)
+
+	return userID, nil
+}
+
+// GetFileContent return the contents of the latest revision of the given file in the database
+// An error is returned if no file with the given id exists.
+func (db *Database) GetFileContent(id int) (obj.File, error) {
+	var f obj.File
+
+	conn, err := sqlite3.Open(db.Path)
+	defer conn.Close()
+	if err != nil {
+		return f, err
+	}
+
+	rows, err := conn.Query("SELECT name, data, owner FROM files WHERE id=?", id)
+	if rows != nil {
+		defer rows.Close()
+	}
+	if err != nil {
+		return f, err
+	}
+	var data string
+	var user string
+	var name string
+	rows.Scan(&name, &data, &user)
+
+	f.Content = data
+	f.Name = name
+	f.User = user
+	f.ID = id
+
+	return f, nil
+}
+
+// GetChangesSinceRevision returns an array of all changes made to the given file after the given revision number.
+func (db *Database) GetChangesSinceRevision(id int, revisionNumber int) ([]obj.Change, error) {
+	conn, err := sqlite3.Open(db.Path)
+	defer conn.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	revs, err := conn.Query("SELECT DISTINCT changes.file, rev_number, position, character FROM revisions JOIN changes WHERE changes.file = ? AND number > ?", id, revisionNumber)
```

Comment:
Line breaks here too would make things a bit cleaner.
Something like:
```suggestion
	revs, err := conn.Query(`
	 SELECT DISTINCT changes.file, rev_number, position, character
	 FROM revisions JOIN changes
	 WHERE changes.file = ? AND number > ?`, id, revisionNumber)
```

If you do this, you might have to play around a bit before the compiler is happy. Note that the [back-tick allows multi-line strings](https://stackoverflow.com/questions/7933460/how-do-you-write-multiline-strings-in-go).

Reply:
**Fixed.** Good point, I did leave some of the super short queries on one line.